man/man1/ec2publishimg.1
.\" Process this file with
.\" groff -man -Tascii ec2publishimg.1
.\"
.TH ec2publishimg.1
.SH NAME
ec2publishimg \- Modify the visibility permissions for an AMI
.SH SYNOPSIS
.B ec2publishimg [option]
.SH DESCRIPTION
.B ec2publishimg
Set the visibility of an AMI to allow others to use the image or to make the
image private, i.e. only available to the account owner.
.SH OPTIONS
.IP "-a --account ACCOUNT_NAME"
Specifies the account to use to connect to EC2. The account is specified
in the config file
.IR ~/.ec2utils.conf .
The account is specified as a section with [account-ACCOUNT_NAME] delimiter.
The options of the sections are:
.IR access_key_id ,
.IR secret_access_key ,
.I ssh_key_name
and
.IR ssh_private_key .
These allow the program to connect to EC2.
If the access_key_id and/or secret_access_key are not found in
.IR ~/.ec2utils.conf,
the search will use the ACCOUNT_NAME to look for a matching section
[profile ACCOUNT_NAME] in
.IR ~/.aws/config
or [ACCOUNT_NAME] in
.IR ~/.aws/credentials.
.IP "--access-id AWS_ACCESS_KEY"
Specifies the AWS access key and overrides the value given for the
.I account
with the
.I access_key_id
in the configuration file.
.IP "--allow-copy"
Set the image copy permissions. The option supports the keyword
.I image
to allow those that the image is shared with to copy it; the keyword
.I none
which does not allow copy access and is the default behavior. The option
allows the specification of an AWS account number or a comma separated list
with no white space to specify multiple account numbers to allow those
accounts to copy the image.
.IP "-n --dry-run"
The program will not perform any action. It will provide information on
.I stdout
about the actions it would perform.
.IP "-f --file CONFIG_FILE"
Specifies the configuration file to use. The default is
.IR ~/.ec2utils.conf .
.IP "--image-id AMI_ID"
Specify the AMI ID of the image to be published. This option is
mutually exclusive with
.IR --image-name ,
.IR --image-name-frag ,
and
.IR --image-name-match .
.IP "--image-name IMAGE_NAME"
Specify the name of the image to be published. The program will look for
an exact match of the name. This option is mutually exclusive with
.IR --image-id ,
.IR --image-name-frag ,
and
.I --image-name-match
options.
.IP "--image-name-frag IMAGE_NAME_FRAGMENT"
Specify a section of an image name for the image(s) to be published. Every
image that matches the name fragment will be published. This
option is mutually exclusive with the
.IR --image-id ,
.IR --image-name-name ,
and
.I --image-name-match
options.
.IP "--image-name-match REGEX"
Specify a regular expression to match an image name. Every image matching the
regular expression will be published. This option is mutually
exclusive with the
.IR --image-id ,
.IR --image-name-name ,
and
.I --image-name-frag
options.
.IP "-r --regions EC2_REGIONS"
A comma separated list of Amazon EC2 regions, or a single region. If no
region argument is specified all EC2 connected regions will be processed.
EC2 disconnected regions have different API keys and thus the specified
.I account
would not work with the disconnected regions. If a disconnected region is to
be processed specify the region explicitly on the command line, and only the
region of interest along with the matching
.IR account .
.IP "-s --secret-key AWS_SECRET_KEY"
Specifies the AWS secret access key and overrides the value given for the
.I account
with the
.I secret_access_key
in the configuration file.
.IP "--share-with SHARE"
Specify the scope of the image publishing/sharing. The option supports the
keyword
.I all
to set the image to public, the keyword
.I none
to set the image private, or expects a AWS account number to share the image
with a specific account. Use a comma separated list with no white space to
specify multiple account numbers. By default the selected image will be
published, i.e.
.I all
is the default value.
.IP "--verbose"
Print extra output about the operations performed to STDOUT.
.IP "--version"
Print the version of he program
.SH EXAMPLE
ec2publishimg --account example --image-name-match production-v2 --share-with all

Will set all images in all connected regions that match the 
.I production-v2
regular expression as public.
.SH AUTHOR
Robert Schweikert (rjschwei@suse.com)


lib/ec2imgutils/ec2publishimg.py
# Copyright 2020 SUSE LLC
#
# This file is part of ec2imgutils
#
# ec2imgutils is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ec2imgutils is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with ec2publishimg. If not, see <http://www.gnu.org/licenses/>.

import logging

import ec2imgutils.ec2utils as utils
from ec2imgutils.ec2imgutils import EC2ImgUtils
from ec2imgutils.ec2imgutilsExceptions import EC2PublishImgException


class EC2PublishImage(EC2ImgUtils):
    """Publish/share EC2 image(s). Will also hide the image(s)"""

    # --------------------------------------------------------------------
    def __init__(
            self,
            access_key=None,
            allow_copy='none',
            image_id=None,
            image_name=None,
            image_name_fragment=None,
            image_name_match=None,
            secret_key=None,
            visibility='all',
            log_level=logging.INFO,
            log_callback=None
    ):
        EC2ImgUtils.__init__(
            self,
            log_level=log_level,
            log_callback=log_callback
        )

        self.access_key = access_key
        self.image_id = image_id
        self.image_name = image_name
        self.image_name_fragment = image_name_fragment
        self.image_name_match = image_name_match
        self.secret_key = secret_key
        self.visibility = visibility

        if self.visibility == 'all':
            self.publish_msg = '\tPublished: %s\t\t%s'
        elif self.visibility == 'none':
            self.publish_msg = '\tSet to private: %s\t\t%s'
        else:
            self.publish_msg = '\tShared: %s\t\t%s\t with: %s'

        if allow_copy == 'image':
            self.allow_copy = self.visibility
        else:
            self.allow_copy = allow_copy

    # --------------------------------------------------------------------
    def _get_images(self):
        """Return a list of images that match the filter criteria"""
        self._connect()
        owned_images = self._get_owned_images()
        if self.image_id:
            return utils.find_images_by_id(owned_images, self.image_id)
        elif self.image_name:
            return utils.find_images_by_name(
                owned_images,
                self.image_name,
                self.log
            )
        elif self.image_name_fragment:
            return utils.find_images_by_name_fragment(
                owned_images,
                self.image_name_fragment,
                self.log
            )
        elif self.image_name_match:
            try:
                return utils.find_images_by_name_regex_match(
                    owned_images,
                    self.image_name_match,
                    self.log
                )
            except Exception:
                msg = 'Unable to complie regular expression "%s"'
                msg = msg % self.image_name_match
                raise EC2PublishImgException(msg)

    # --------------------------------------------------------------------
    def _get_snapshot_ids_for_image(self, image):
        """Return the snapshot ID for a given image"""
        image_data = self._connect().describe_images(
            ImageIds=[image['ImageId']])['Images']
        block_device_maps = image_data[0]['BlockDeviceMappings']
        snapshot_ids = []
        for block_map in block_device_maps:
            snapshot_ids.append(block_map['Ebs']['SnapshotId'])

        return snapshot_ids

    # --------------------------------------------------------------------
    def _share_snapshot(self, image):
        """Provide permission to copy the underlying snapshot"""

        snapshot_ids = self._get_snapshot_ids_for_image(image)
        for snapshot_id in snapshot_ids:
            if self.allow_copy == 'all':
                self._connect().modify_snapshot_attribute(
                    SnapshotId=snapshot_id,
                    Attribute='createVolumePermission',
                    OperationType='add',
                    GroupNames=['all']
                )
            elif self.allow_copy != 'none':
                self._connect().modify_snapshot_attribute(
                    SnapshotId=snapshot_id,
                    Attribute='createVolumePermission',
                    OperationType='add',
                    UserIds=self.allow_copy.split(',')
                )

    # --------------------------------------------------------------------
    def _print_image_info(self, image, log_callback):
        """Print a message about the image that would be modified"""
        if self.visibility == 'all' or self.visibility == 'none':
            log_callback(
                self.publish_msg % (image['ImageId'], image['Name'])
            )
        else:
            log_callback(self.publish_msg % (
                        image['ImageId'],
                        image['Name'],
                        self.visibility
            ))

    # --------------------------------------------------------------------
    def print_publish_info(self):
        """Print information about images that would be published"""
        images = self._get_images()
        for image in images:
            self._print_image_info(image, log_callback=self.log.info)

    # --------------------------------------------------------------------
    def publish_images(self):
        """Publish the matching image(s)"""
        images = self._get_images()

        for image in images:
            skip_state = ['invalid', 'deregistered', 'transient', 'fail']
            if image['State'] == 'pending':
                waiter = self._connect().get_waiter('image_available')
                self.log.info("Current state of image %s is %s. Waiting up "
                              "to 10 minutes for it to become 'available'.",
                              image['ImageId'], image['State'])
                wait_status = 1
                try:
                    wait_status = waiter.wait(
                        ImageIds=[image['ImageId']],
                        Filters=[
                            {
                                'Name': 'state',
                                'Values': ['available']
                            }
                        ],
                        WaiterConfig={
                            'Delay': 15,
                            'MaxAttempts': 40
                        }
                    )
                except Exception:
                    wait_status = 1

                if wait_status:
                    self.log.info("Skipping image %s as it failed to become"
                                  "'available'", image['ImageId'])
                    continue
                else:
                    self.log.info("Image %s is now 'available'.",
                                  image['ImageId'])

            elif image['State'] in skip_state:
                self.log.info("Skipping image %s as the state is %s"
                              " and is not able to be published.",
                              image['ImageId'], image['State'])
                continue

            if self.visibility == 'all':
                self._connect().modify_image_attribute(
                    ImageId=image['ImageId'],
                    Attribute='launchPermission',
                    OperationType='add',
                    UserGroups=['all']
                )
                if self.allow_copy != 'none':
                    self._share_snapshot(image)
            elif self.visibility == 'none':
                launch_attributes = self._connect().describe_image_attribute(
                    ImageId=image['ImageId'],
                    Attribute='launchPermission'
                )['LaunchPermissions']
                launch_permission = {
                    'Remove': launch_attributes
                }
                if not launch_attributes:
                    msg = '\tImage with ID: %s  ' % image['ImageId']
                    msg += 'is already private, nothing to do'
                    self.log.info(msg)
                    continue
                self._connect().modify_image_attribute(
                    ImageId=image['ImageId'],
                    LaunchPermission=launch_permission
                )
                snapshot_ids = self._get_snapshot_ids_for_image(image)
                for snapshot_id in snapshot_ids:
                    snapshot_attrs = (
                        self._connect().describe_snapshot_attribute(
                            SnapshotId=snapshot_id,
                            Attribute='createVolumePermission'
                        )['CreateVolumePermissions']
                    )
                    if not snapshot_attrs:
                        continue
                    snapshot_permission = {
                        'Remove': snapshot_attrs
                    }
                    self._connect().modify_snapshot_attribute(
                        SnapshotId=snapshot_id,
                        CreateVolumePermission=snapshot_permission
                    )
            else:
                self._connect().modify_image_attribute(
                    ImageId=image['ImageId'],
                    Attribute='launchPermission',
                    OperationType='add',
                    UserIds=self.visibility.split(',')
                )
                if self.allow_copy:
                    self._share_snapshot(image)

            self._print_image_info(image, log_callback=self.log.debug)
